karapace/protobuf/schema.py

```diff
@@ -1,14 +1,168 @@
+        '\'': "\\'",
+        '\"': '\\"',
+        '\\': '\\\\'
+    }
+    result: str = ""
+    for char in text:
+        try:
+            result += escape_dict[char]
+        except KeyError:
+            result += char
+    return result
+
+
+def message_element_string(element: MessageElement) -> str:
+    result: list = list()
+    append_documentation(result, element.documentation)
+    result.append(f"message {element.name} {{")
+    if element.reserveds:
+        result.append("\n")
+        for reserved in element.reserveds:
+            append_indented(result, reserved.to_schema())
+
+    if element.options:
+        result.append("\n")
+        for option in element.options:
+            append_indented(result, option_element_string(option))
+
+    if element.fields:
+        result.append("\n")
+        for field in element.fields:
+            append_indented(result, field.to_schema())
+
+    if element.one_ofs:
+        result.append("\n")
+        for one_of in element.one_ofs:
+            append_indented(result, one_of.to_schema())
+
+    if element.groups:
+        result.append("\n")
+        for group in element.groups:
+            append_indented(result, group.to_schema())
+
+    if element.extensions:
+        result.append("\n")
+        for extension in element.extensions:
+            append_indented(result, extension.to_schema())
+
+    if element.nested_types:
+        result.append("\n")
+        for nested_type in element.nested_types:
+            if isinstance(nested_type, MessageElement):
+                append_indented(result, message_element_string(nested_type))
+
+        for nested_type in element.nested_types:
+            if isinstance(nested_type, EnumElement):
+                append_indented(result, enum_element_string(nested_type))
+
+    result.append("}\n")
+    return "".join(result)
+
+
+def enum_element_string(element: EnumElement) -> str:
+    return element.to_schema()
+
+
+def option_element_string(option: OptionElement):
+    result: str
+    if option.kind == OptionElement.Kind.STRING:
+        name: str
+        if option.is_parenthesized:
+            name = f"({option.name})"
+        else:
+            name = option.name
+        value = add_slashes(str(option.value))
+        result = f"{name} = \"{value}\""
+    else:
+        result = option.to_schema()
+
+    return f"option {result};\n"
 
 
 class ProtobufSchema:
+    DEFAULT_LOCATION = Location.get("")
+
     def __init__(self, schema: str):
-        self.schema = schema
+        self.dirty = schema
+        self.cache_string = ""
+        self.schema = ProtoParser.parse(self.DEFAULT_LOCATION, schema)
 
     def __str__(self) -> str:
-        return self.schema
+        if not self.cache_string:
+            self.cache_string = self.to_schema()
+        log.warning("CACHE_STRING:%s", self.cache_string)
+        return self.cache_string
 
     def to_json(self) -> str:
-        return self.schema
+        return self.to_schema()
```

Comment:
if we don't actually support for converting to json, should we just raise an error here or remove the function altogether?